apps/prairielearn/src/lib/editors.js

```diff
@@ -677,88 +661,52 @@ class AssessmentCopyEditor extends Editor {
-        (callback) => {
-          debug(`Generate TID and Title`);
-          let names = this.getNamesForCopy(
-            this.assessment.tid,
-            this.oldNamesShort,
-            this.assessment.title,
-            this.oldNamesLong
-          );
-          this.tid = names.shortName;
-          this.assessmentTitle = names.longName;
-          this.assessmentPath = path.join(assessmentsPath, this.tid);
-          this.pathsToAdd = [this.assessmentPath];
-          this.commitMessage = `${this.course_instance.short_name}: copy assessment ${this.assessment.tid} to ${this.tid}`;
-          callback(null);
-        },
-        (callback) => {
-          const fromPath = path.join(assessmentsPath, this.assessment.tid);
-          const toPath = this.assessmentPath;
-          debug(`Copy template\n from ${fromPath}\n to ${toPath}`);
-          fs.copy(fromPath, toPath, { overwrite: false, errorOnExist: true }, (err) => {
-            if (ERR(err, callback)) return;
-            callback(null);
-          });
-        },
-        (callback) => {
-          debug(`Read infoAssessment.json`);
-          fs.readJson(path.join(this.assessmentPath, 'infoAssessment.json'), (err, infoJson) => {
-            if (ERR(err, callback)) return;
-            callback(null, infoJson);
-          });
-        },
-        (infoJson, callback) => {
-          debug(`Write infoAssessment.json with new title and uuid`);
-          infoJson.title = this.assessmentTitle;
-          this.uuid = uuidv4(); // <-- store uuid so we can find the new assessment in the DB
-          infoJson.uuid = this.uuid;
-          fs.writeJson(
-            path.join(this.assessmentPath, 'infoAssessment.json'),
-            infoJson,
-            { spaces: 4 },
-            (err) => {
-              if (ERR(err, callback)) return;
-              callback(null);
-            }
-          );
-        },
-      ],
-      (err) => {
-        if (ERR(err, callback)) return;
-        callback(null);
-      }
+
+    debug('Get all existing long names');
+    const result = await sqldb.queryAsync(sql.select_assessments_with_course_instance, {
+      course_instance_id: this.course_instance.id,
+    });
+    const oldNamesLong = _.map(result.rows, 'title');
+
+    debug('Get all existing short names');
+    const oldNamesShort = await this.getExistingShortNames(assessmentsPath, 'infoAssessment.json');
+
+    debug(`Generate TID and Title`);
+    let names = this.getNamesForCopy(
+      this.assessment.tid,
+      oldNamesShort,
+      this.assessment.title,
+      oldNamesLong
     );
+    const tid = names.shortName;
+    const assessmentTitle = names.longName;
+    const assessmentPath = path.join(assessmentsPath, tid);
+    this.pathsToAdd = [assessmentPath];
+    this.commitMessage = `${this.course_instance.short_name}: copy assessment ${this.assessment.tid} to ${tid}`;
+
+    const fromPath = path.join(assessmentsPath, this.assessment.tid);
+    const toPath = assessmentPath;
```

Comment:
I'm not tracking the details, but I assume it's ok that we lost the `this` here?

Reply:
Yep! Previously some of these values were being written to `this` so that they could be accessed in the various `async.series` callbacks. I rewrote them as plain old local variables, which should be much easier to reason about.